libraries/ms-drivers/src/stm32f0xx/mcp3427_adc.c

```diff
@@ -0,0 +1,196 @@
+#include "mcp3427_adc.h"
+#include "fsm.h"
+#include "log.h"
+#include "mcp3427_adc_defs.h"
+#include "soft_timer.h"
+
+// Note: we always read from both channels before reporting, taking 100ms between reports. If timing
+// is an issue, an optimization for boards (i.e. solar) using only one channel is to only read that
+// channel, so reports take 50ms.
+
+#define MCP3427_FSM_NAME "MCP3427 FSM"
+
+#define NUM_MCP3427_CHIP_IDS (1 << 4)
+
+// A lookup table of MCP3427 chip IDs (see |prv_get_chip_identifier|) to their storages,
+// used to automagically direct events to the correct storage in |mcp3427_process_event|.
+// This saves having to pass each event to every MCP3427.
+static Mcp3427Storage *s_id_to_storage_cache[NUM_MCP3427_CHIP_IDS] = { 0 };
+
+FSM_DECLARE_STATE(channel_1_trigger);
+FSM_DECLARE_STATE(channel_1_readback);
+FSM_DECLARE_STATE(channel_2_trigger);
+FSM_DECLARE_STATE(channel_2_readback);
+
+FSM_STATE_TRANSITION(channel_1_trigger) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_ready_event, channel_1_readback);
+}
+
+FSM_STATE_TRANSITION(channel_1_readback) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_trigger_event, channel_2_trigger);
+}
+
+FSM_STATE_TRANSITION(channel_2_trigger) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_ready_event, channel_2_readback);
+}
+
+FSM_STATE_TRANSITION(channel_2_readback) {
+  Mcp3427Storage *storage = fsm->context;
+  FSM_ADD_TRANSITION(storage->data_trigger_event, channel_1_trigger);
+}
+
+static uint8_t prv_get_chip_identifier(Mcp3427Storage *storage) {
+  // Used to gate events we raised to only this MCP3427.
+  // ID is 4 bits, with the "base address" (the 3 least significant bits of the I2C address) as the
+  // most significant bits and the I2C port encoded as the LSB.
+  uint8_t base_addr = storage->addr ^ (MCP3427_DEVICE_CODE << 3);
+  return (base_addr << 1) | (storage->port == I2C_PORT_1 ? 0 : 1);
```

Comment:
is there any reason you're cramming this into a `uint8_t`? couldn't you just define a chip identifier struct? I think this is fine, but I feel like best practice is probably to not have complex expressions like this since they tend to be most error prone.

Reply:
I'm cramming it into the rightmost 4 bits of a `uint8_t` because I'm using the chip ID as an index for `s_id_to_storage_cache`, and I'm also using it as the data for the events so that we can direct them to the appropriate storage. This would be a lot harder to do with a struct.

Reply:
Fair enough. Something we have in centre console is a 
```
typedef union FaultReason {
    struct {
        uint8_t area;
        uint8_t reason;
    } fields;
    uint16_t raw;
} FaultReason;
```
But To be honest that's not really any easier to work with than packing into an int since it tends to add a lot of lines wherever you declare them. For events you pass in the `fault.raw`, then when unpacking them you do `FaultReason fault = { .raw = e.data };` then access it via `fault.fields.reason` etc.